### PR TITLE
Added basic WebSocket support in the embedded server

### PR DESCRIPTION
**Pull request by nacx**
This PR is based on the comments in https://github.com/perwendel/spark/issues/196.
Adds basic support for WebSockets in the embedded server.

### CODE REVIEW

File: src/main/java/spark/SparkBase.java [hunk 1]
```diff
@@ -358,7 +388,7 @@ protected static void addFilter(String httpMethod, FilterImpl filter) {
                                                    + "'", filter.getAcceptType(), filter);
     }
 
-    private static synchronized void init() {
+    public static synchronized void init() {
```
Comment (nacx) [inline src/main/java/spark/SparkBase.java]: Note this has been made public to allow calling N times the `webSocket` method, and allow to start Spark serving only WebSockets. If no routes are added, users should be able to manually call `init()` after having declared the websocket handlers.

**Comment by perwendel** [comment]
Tx, will have a look at this after my vacation (the week after next)

**Comment by joakime** [comment]
This can be better handled with use of [`org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter`](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/websocket/server/WebSocketUpgradeFilter.html) in Jetty 9.2.+

Use its [`.addMapping(PathSpec, WebSocketCreator)`](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/websocket/server/WebSocketUpgradeFilter.html#addMapping-org.eclipse.jetty.websocket.server.pathmap.PathSpec-org.eclipse.jetty.websocket.servlet.WebSocketCreator-) method.

**Comment by perwendel** [comment]
I've merged the PR that bumps the jetty version to 9.2.11.v20150529. Do you (@nacx) modify this PR with the tip from @joakime?

Btw, this PR looks good!

**Comment by nacx** [comment]
Sure! I'll rebase the branch and update the PR.

**Comment by nacx** [comment]
@perwendel @joakime I've rebased and squashed the commits into a single one and addressed the comments. Please have a look!
Using the `WebSocketUpgradeFilter` has made possible to get rid of the `WebSocketServlet` so it is no longer exposed to users. They just need to pass the handler class properly annotated without having to extend the servlet.

File: src/main/java/spark/SparkBase.java [hunk 2]
```diff
@@ -241,6 +249,38 @@ public static synchronized void externalStaticFileLocation(String externalFolder
...
+        webSocketHandlers.put(path, handler);
+    }
+
+    /**
+     * Sets the max idle timeout in milliseconds for WebSocket connections.
+     *
+     * @param timeoutMillis The max idle timeout in milliseconds.
+     */
+    public static void webSocketIdleTimeoutMillis(int timeoutMillis) {
+	webSocketIdleTimeoutMillis = Optional.of(timeoutMillis);
```
Comment (perwendel) [inline src/main/java/spark/SparkBase.java]: no if (initialized) check and exception if already initialized?

File: src/main/java/spark/SparkBase.java [hunk 3]
```diff
@@ -241,6 +249,38 @@ public static synchronized void externalStaticFileLocation(String externalFolder
...
 
     /**
+     * Maps the given path to the given WebSocket handler.
+     * <p>
+     * This is currently only available in the embedded server mode.
+     *
+     * @param path the WebSocket path.
+     * @param handler the handler class that will manage the WebSocket connection to the given path.
+     */
+    public static synchronized void webSocket(String path, Class<?> handler) {
```
Comment (perwendel) [inline src/main/java/spark/SparkBase.java]: are the no restrictions on class type here (essentially in WebSocketCreatorFactory.create(webSocketHandlers.get(path))?

Comment (nacx) [inline src/main/java/spark/SparkBase.java]: WebSocket handler classes can be arbitrary classes. Users can:

* Implement the `WebSocketListener` interface (or extend the convenient `WebSocketAdapter`).
* Use an arbitrary class and use the websocket annotations, as seen in [the examples](https://github.com/nacx/spark/blob/ws/src/test/java/spark/examples/websocket/EchoWebSocket.java).

That's why I left it wihtout restrictions, but I'll add a check to require the annotation or the listener interface.

**Comment by perwendel** [comment]
@nacx Looks good. One comment and one question.

**Comment by nacx** [comment]
Will add the checks for the initialized and the websocket handlers. I'm on it now. Thanks for reviewing!

File: src/main/java/spark/SparkBase.java (hunk 3 above)
Comment (perwendel) [inline src/main/java/spark/SparkBase.java]: Ok, do what you find is best regarding the handlers.

**Comment by nacx** [comment]
@perwendel Done. I've added a new commit to facilitate the review. It it looks good I'll squash the commit for a cleaner merge if needed.

**Comment by perwendel** [comment]
:+1: 
Good stuff man! Thanks a lot!

**Comment by nacx** [comment]
Thanks! Should #196 and #65 be marked as fixed now?

Also, I've seen the formatting follow-up commit. Is there any style guide or config I can import for future contributions?

**Comment by vemacs** [comment]
@nacx I'm unable to get even the provided examples working on a standalone server. I can confirm that they're being registered (see https://github.com/perwendel/spark/issues/345#issuecomment-137273915), but any attempts to connect to them using a standard browser fails, and I get `The requested route [/echo] has not been mapped in Spark` on the server console.

**Pull request merged**